[1066] Update right to work or study flow

avinhurry opened the pull request:
## Context

Some users have a difficult time understanding what they are being asked when presented with an 'immigration status' question. This PR aims to address the issues found during the last round of research.

## Changes proposed in this pull request

See commits

## Guidance to review

## Link to Trello card

https://trello.com/c/b8KPqnhE/1066-apply-visa-immigration-status

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

avinhurry commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Apply: Visa/immigration status](https://trello.com/c/b8KPqnhE/1066-apply-visa-immigration-status)

tomas-stefano commented on app/helpers/candidate_details_helper.rb:
```diff
@@ -0,0 +1,5 @@
+  def includes_eu_eea_swiss?(nationalities)
+    EU_EEA_SWISS_COUNTRY_CODES.intersect?(nationalities.map { |name| NATIONALITIES_BY_NAME[name] })
+  end
+end
```
@JR-G  @inulty-dfe I wanna hear your opinions about helpers and form objects. Should we have a concerns folder in app/forms or move everything to app/helpers? @avinhurry can you investigate some other modules that is used in app/forms that were moved to app/helpers please?

inulty-dfe replied:
Looks like `Nationalities` is a module.

A look up interface would be useful here. We'd be able to enquire it's methods rather than them all being unrelated constants in `config/initializers/nationalities.rb`

avinhurry replied:
Let's consider this as a separate piece of work as this PR is already quite big and we could look at using a lookup for the other constants in nationalities.

JR-G commented:
Looking good! I've just tested the API with postman and the `uk_residency_status` field is returning null:
![image](https://github.com/DFE-Digital/apply-for-teacher-training/assets/47917431/31f3cb78-3950-4006-9ae9-b9c2896bf7b9)
vs main:
![image](https://github.com/DFE-Digital/apply-for-teacher-training/assets/47917431/bf879d40-de8a-4089-9e86-b82dbc2e5c4a)

JR-G commented:
The value also isn't rendering for me on the support UI:
<img width="561" alt="image" src="https://github.com/DFE-Digital/apply-for-teacher-training/assets/47917431/85ccf1fc-80f8-489d-b455-fd77e109bc5d">
and I think with the change link we'd also need to render the same options as the flow?

avinhurry commented:
> Looking good! I've just tested the API with postman and the `uk_residency_status` field is returning null: ![image](https://private-user-images.githubusercontent.com/47917431/299704014-31f3cb78-3950-4006-9ae9-b9c2896bf7b9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYyMjc2OTQsIm5iZiI6MTcwNjIyNzM5NCwicGF0aCI6Ii80NzkxNzQzMS8yOTk3MDQwMTQtMzFmM2NiNzgtMzk1MC00MDA2LTlhZTktYjljMjg5NmJmN2I5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI2VDAwMDMxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjNDk5OWFkOTFkZjdlY2JmZDhiYjYwNGUzZDcwOGZjNDQ2ZWQyNzEwMzM4YzNlNzhhYTNjOTQ4MDFhZjM0ZTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.nA6MHTyzwj4nd0ixaqdBZRnqQvqqOEBpBWjpz40-iw8) vs main: ![image](https://private-user-images.githubusercontent.com/47917431/299704094-bf879d40-de8a-4089-9e86-b82dbc2e5c4a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYyMjc2OTQsIm5iZiI6MTcwNjIyNzM5NCwicGF0aCI6Ii80NzkxNzQzMS8yOTk3MDQwOTQtYmY4NzlkNDAtZGU4YS00MDg5LTllODYtYjgyZGJjMmU1YzRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI2VDAwMDMxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIyMWQyMjUyZmI0OGRmNDczMDYxNDFhODQ0ZGJkMDhlMGE5MzNlMzYzZWUwZWY2MDM3ZThiMGE2NWMyNTE0ZjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NcI1eevR2_shpT4TySB-NeNAwdQR5XQw8OUE_jAw0ig)
> 
> I think `right_to_work_or_study_details` might be the culprit:
> 
> ```
> class FormatResidencyDetailsService
>   def initialize(application_form:)
>     @application_form = application_form
>   end
> 
>   def residency_details_value
>     if @application_form.eu_settled? || @application_form.eu_pre_settled?
>       I18n.t("application_form.personal_details.immigration_status.values.#{@application_form.immigration_status}")
>     else
>       @application_form.right_to_work_or_study_details
>     end
>   end
> end
> ```

Nice one @JR-G, thanks! 

I think i've addressed those issues now. Would you mind giving the API another test on postman please?

I think the test that's failing in this PR is failing on main.

JR-G commented:
> > Looking good! I've just tested the API with postman and the `uk_residency_status` field is returning null: ![image](https://private-user-images.githubusercontent.com/47917431/299704014-31f3cb78-3950-4006-9ae9-b9c2896bf7b9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYyMjc2OTQsIm5iZiI6MTcwNjIyNzM5NCwicGF0aCI6Ii80NzkxNzQzMS8yOTk3MDQwMTQtMzFmM2NiNzgtMzk1MC00MDA2LTlhZTktYjljMjg5NmJmN2I5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI2VDAwMDMxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjNDk5OWFkOTFkZjdlY2JmZDhiYjYwNGUzZDcwOGZjNDQ2ZWQyNzEwMzM4YzNlNzhhYTNjOTQ4MDFhZjM0ZTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.nA6MHTyzwj4nd0ixaqdBZRnqQvqqOEBpBWjpz40-iw8) vs main: ![image](https://private-user-images.githubusercontent.com/47917431/299704094-bf879d40-de8a-4089-9e86-b82dbc2e5c4a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYyMjc2OTQsIm5iZiI6MTcwNjIyNzM5NCwicGF0aCI6Ii80NzkxNzQzMS8yOTk3MDQwOTQtYmY4NzlkNDAtZGU4YS00MDg5LTllODYtYjgyZGJjMmU1YzRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI2VDAwMDMxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIyMWQyMjUyZmI0OGRmNDczMDYxNDFhODQ0ZGJkMDhlMGE5MzNlMzYzZWUwZWY2MDM3ZThiMGE2NWMyNTE0ZjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NcI1eevR2_shpT4TySB-NeNAwdQR5XQw8OUE_jAw0ig)
> > I think `right_to_work_or_study_details` might be the culprit:
> > ```
> > class FormatResidencyDetailsService
> >   def initialize(application_form:)
> >     @application_form = application_form
> >   end
> > 
> >   def residency_details_value
> >     if @application_form.eu_settled? || @application_form.eu_pre_settled?
> >       I18n.t("application_form.personal_details.immigration_status.values.#{@application_form.immigration_status}")
> >     else
> >       @application_form.right_to_work_or_study_details
> >     end
> >   end
> > end
> > ```
> 
> Nice one @JR-G, thanks!
> 
> I think i've addressed those issues now. Would you mind giving the API another test on postman please?

Nice! API is working :) 
<img width="642" alt="image" src="https://github.com/DFE-Digital/apply-for-teacher-training/assets/47917431/f15b202b-2253-4e93-b117-07c3b888bb44">

vassyz commented on spec/forms/candidate_interface/immigration_status_form_spec.rb:
```diff
@@ -22,6 +22,33 @@
+
+        it 'is valid' do
+          expect(form.valid?).to be true
+          expect(form.errors[:right_to_work_or_study_details]).to be_empty
```
This is a bit redundant if `valid?` is `true.

vassyz commented on spec/system/candidate_interface/entering_details/candidate_choosing_visa_or_immigration_status_non_eu_spec.rb:
```diff
@@ -0,0 +1,106 @@
+require 'rails_helper'
+
+RSpec.describe 'Choosing visa or immigration status' do
+  scenario 'non eu candidate who and has right to work' do
```
Some typo here.

vassyz commented on spec/system/candidate_interface/entering_details/candidate_choosing_visa_or_immigration_status_non_eu_spec.rb:
```diff
@@ -0,0 +1,106 @@
+    when_click_change_immigration_status
+    and_i_choose_the_visa('Refugee Status')
+    and_i_click_save_and_continue
+    then_i_should_see_the_correct_visa_in_the_summary('Refugee Status')
```
I'm not sure if testing all the values should be done in a system spec.

avinhurry replied:
yeah, I wasn't keen either but I couldn't think of a better way to properly test it. Do you have any suggestions?

vassyz replied:
You already wrote the specs, so might as well leave it now. I see system specs as a tool for testing interactions, not the data. And if you add another `immigration_status` value in the model, this spec will not fail so the dev might forget to add the scenario.

avinhurry commented on app/models/application_form.rb:
```diff
@@ -196,6 +210,12 @@ def submitted?
     submitted_at.present?
   end
 
+  def defined_immigration_status?
```
I was trying to keep `spec/presenters/register_api/single_application_presenter_spec.rb` from failing, that version causes it to fail 😞

vassyz replied:
It's because it calls `immigration_status`. But your version technically does the same. You have to add `immigration_status` to `PUBLISHED_FIELDS` but I'm not entirely sure what the implications of that are. CC @JR-G @tomas-stefano

tomas-stefano replied:
Going into published fields means that will update all application choices if there is a update in the field, therefore they will bump in the order those choices shows into the API. But the moment you add to published fields the specs will verify if the presenter is query for that field and fails otherwise. Is that the desired output? @JR-G if so I'd say @vassyz suggestion is good.
If the above statement is true you can also do `immigration_status.in?(ApplicationForm.immigration_statuses.keys - %w[other])`

JR-G replied:
That's right – but I think in this scenario that's an acceptable change, it would make sense to have it bumped if there immigration status has changed

inulty-dfe commented on app/views/candidate_interface/personal_details/immigration_status/edit.html.erb:
```diff
@@ -1,4 +1,5 @@
-<% content_for :title, title_with_error_prefix(t('page_titles.immigration_status'), @form.errors.any?) %>
+<% content_for :title, title_with_error_prefix(t("page_titles.#{@form.eu_nationality? ? 'immigration_status' : 'visa_or_immigration_status'}"), @form.errors.any?) %>
```
This pattern appears again. Can we improve on this?

avinhurry replied:
https://github.com/DFE-Digital/apply-for-teacher-training/pull/9012/commits/b0dd52986d98f64b073f8d0495c71f2ca1b86d5a

inulty-dfe commented on app/views/support_interface/application_forms/immigration_status/edit.html.erb:
```diff
@@ -8,29 +8,42 @@
+      <h1 class="govuk-heading-l"><%= t('support_interface.page_titles.visa_or_immigration_status') %></h1>
+        <%= f.govuk_radio_buttons_fieldset :right_to_work_or_study_details,
+        legend: { text: 'Select your visa or immigration status' } do %>
+          <%= f.govuk_radio_button :immigration_status, :indefinite_leave_to_remain_in_the_uk, label: { text: t('support_interface.edit_immigration_status.values.indefinite_leave_to_remain_in_the_uk') } %>
```
Same comment about the `scope` approach to reduce the repetition and make the markup easier to scan.

avinhurry replied:
https://github.com/DFE-Digital/apply-for-teacher-training/pull/9012/commits/c23d80bfbb7924618fd4ddb95634838f14b8d1c4

inulty-dfe commented on spec/forms/candidate_interface/immigration_status_form_spec.rb:
```diff
@@ -13,7 +13,7 @@
     context 'when the immigration status is nil' do
       it 'validation keeps immigration_status as `nil`' do
-        form.valid?
+        expect(form.valid?).to be false
```
```ruby
expect(form).not_to be_valid
```

avinhurry replied:
No particular reason. I guess it just came more naturally to me.

inulty-dfe commented on spec/system/support_interface/editing_immigration_status_non_eu_spec.rb:
```diff
@@ -0,0 +1,136 @@
+    then_i_should_see_the_correct_visa_in_the_summary('Refugee Status')
+  end
+
+  def and_i_add_an_audit_comment
```
Can you sort these methods in the order they appear in the test? I've started trying to do this when I can remember to.

avinhurry replied:
Interesting. My thoughts on this are: 

- It would be nice to read down the page and see exactly what the test is doing easily in one go.   
- What happens in specs where there are more than one scenario?
- This could also get messy and take time in large files?
- I'm not sure this provides much value for me personally, I always search for the method?


Either way I’ve done it on this occasion here: https://github.com/DFE-Digital/apply-for-teacher-training/pull/9012/commits/22512a69ee5c9d4d346eb342197faeb2321cee26

inulty-dfe reviewed (commented):
Few comments around organisation and readability.

Otherwise looks fine

avinhurry commented on app/presenters/candidate_interface/personal_details_review_presenter.rb:
```diff
@@ -102,19 +103,15 @@ def right_to_work_rows
       if @application_form.immigration_status
         rows << {
-          key: I18n.t('application_form.personal_details.immigration_status.label'),
+          key: I18n.t("application_form.personal_details.#{includes_eu_eea_swiss?(@application_form.nationalities) ? 'immigration_status' : 'visa_or_immigration_status'}.label"),
```
That's a fair challenge. I've updated those here: https://github.com/DFE-Digital/apply-for-teacher-training/pull/9012/commits/b0dd52986d98f64b073f8d0495c71f2ca1b86d5a

avinhurry commented on app/views/candidate_interface/personal_details/immigration_status/_shared_form.html.erb:
```diff
@@ -5,27 +5,38 @@
+  <h1 class="govuk-heading-l"><%= t('page_titles.visa_or_immigration_status') %></h1>
+  <%= f.govuk_radio_buttons_fieldset :right_to_work_or_study_details,
+    legend: { text: 'Select your visa or immigration status' } do %>
+      <%= f.govuk_radio_button :immigration_status, :indefinite_leave_to_remain_in_the_uk, label: { text: t('application_form.personal_details.immigration_status.values.indefinite_leave_to_remain_in_the_uk') } %>
```
Good suggestion, thanks! I was looking to do something like this to help with scanning these views. I didn't know you could do that.

Updated here: https://github.com/DFE-Digital/apply-for-teacher-training/pull/9012/commits/c23d80bfbb7924618fd4ddb95634838f14b8d1c4
